docs/ldapsync.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "MGR\-LDAPSYNC" "1" "2019-10-25" "" ""
.
.SH "NAME"
\fBmgr\-ldapsync\fR \- Utility to synchronise users between LDAP of your choice and the Uyuni server
.
.SH "SYNOPSIS"
\fBmgr\-ldapsync\fR [option]
.
.SH "DESCRIPTION"
\fBmgr\-ldapsync(1)\fR is a program that synchronises LDAP users with the Uyuni server\. It will add users from LDAP, update them if their attributes changed or remove them from Uyuni\.
.
.SH "OPTIONS"
.
.TP
\fB\-c\fR, \fB\-\-config\fR=[value]
Specifiy an optional configuration file\. Default configuration file is: \fB/etc/rhn/ldapsync\.conf\fR
.
.TP
\fB\-o\fR, \fB\-\-overview\fR
Overview or "dry run" mode\. In this case the \fBmgr\-ldapsync\fR will only show you what is going to change, but will not perform any real actions\.
.
.TP
\fB\-s\fR, \fB\-\-sync\fR
Perform an actual synchronisation\.
.
.TP
\fB\-h\fR, \fB\-\-help\fR
Shows help\.
.
.TP
\fB\-v\fR, \fB\-\-version\fR
Shows current version of the \fBmgr\-ldapsync\fR\.
.
.SH "REQUIREMENTS"
LDAP is very flexible and easy to customise\. Because of this, \fBmgr\-ldapsync\fR allows users to adjust the configuration for the existing LDAP schema\. If not specified otherwise, default configuration file is first looked up in \fB/etc/rhn/ldapsync\.conf\fR\.
.
.P
There are certain requirements how to manage users in LDAP so they are synchronised correctly with the Uyuni server:
.
.IP "1." 4
There should be either at least one \fIgroup\fR of object class \fBgroupOfNames\fR or at least one \fIrole\fR of object class \fBorganizationalRole\fR\. In case of Active Directory, \fIgroup\fR object class should be called \fBgroup\fR\.
.
.IP "2." 4
Each group should have at least one attribute \fBmember\fR with a valid DN of an actual user\.
.
.IP "3." 4
In case a groups are not used, a role has to have at least one attribute \fBroleOccupant\fR with the valid DN of an actual user\.
.
.IP "4." 4
There should be an accessible DN for all users in the LDAP database\.
.
.IP "" 0
.
.P
Each user should have the following \fBmandatory\fR attributes:
.
.IP "\(bu" 4
\fBuid\fR
.
.IP "\(bu" 4
\fBcn\fR
.
.IP "\(bu" 4
\fBname\fR or \fBgivenName\fR (optional, if \fBcn\fR has name and second name)
.
.IP "\(bu" 4
\fBsn\fR (optional, if \fBcn\fR has name and second name)
.
.IP "\(bu" 4
\fBmail\fR
.
.IP "" 0
.
.P
\fBIMPORTANT:\fR The \fBmgr\-ldapsync\fR is not expected to work properly, if the requirements above are not met\.
.
.SH "CONFIGURATION"
Configuration file should be a valid YAML file\.
.
.P
It has two main sections at the root:
.
.IP "1." 4
\fBdirectory\fR: Directory describes all the configuration, required by the LDAP server\.
.
.IP "2." 4
\fBrpc\fR: This section describes all the configuration, required by Uyuni server\.
.
.IP "" 0
.
.P
The \fBdirectory\fR section has the following attributes:
.
.TP
\fBudn\fR (string)
A user\'s DN for the directory to connect\. Example: \fBuid=admin,ou=system\fR\.
.
.TP
\fBpassword\fR (string)
LDAP authentication password for the \fBudn\fR above\.
.
.TP
\fBhost\fR (string)
Fully qualified domain name of the LDAP server\.
.
.TP
\fBport\fR (integer, optional)
Port on which LDAP server is running\. By default it is \fB389\fR\.
.
.TP
\fBallusers\fR (string)
DN for all the users subtree\. Example: \fBou=users,dc=example,dc=com\fR\.
.
.P
The \fBdirectory\fR section has also the following directives:
.
.IP "1." 4
\fBfrozen\fR (map, mandatory)\. This is a list of Uyuni user IDs that \fBmgr\-ldapsync\fR should \fIcompletely\fR ignore\. Any user ID, specified in this directive will always exclude any user in the Uyuni server from being affected by LDAP operations\. This is usually the main "static" administrator account or an emergency login\. This directive is mandatory and LDAP sync will refuse to work if you have no at least one frozen user with \fBorg_admin\fR permissions assigned\.
.
.IP "2." 4
\fBgroups\fR or \fBroles\fR map, at least one of them must be present\. \fBEither\fR directive is \fBmandatory\fR to specify, in order to properly manage Uyuni roles\. Both directives has the same structure a list of Uyuni roles, attached to a CN in the LDAP\. See "examples" section below for more details:
.
.IP "" 0
.
.IP "" 4
.
.nf

   roles|groups:
     cn:
       \- role
       \- role
       \- role
       \.\.\.
.
.fi
.
.IP "" 0
.
.P
The \fBrpc\fR section contains all the necessary information for XML\-RPC API of Uyuni server:
.
.TP
\fBurl\fR (string)
URL of the XML\-RPC API for the Uyuni server\. It should contain a schema in order to specify if SSL/TLS is required\.
.
.TP
\fBcheckssl\fR (boolean, optional, default \fBtrue\fR)
A boolean true/false option to allow SSL connection to the URL without proper SSL certificate\.
.
.TP
\fBuser\fR (string)
Username of the Uyuni administrator with the \fBorg_admin\fR or \fBsatellite_admin\fR role\. This is usually the user was created at the very first start\.
.
.TP
\fBpassword\fR (string)
Password for the Uyuni administrator username\.
.
.SH "LIST OF UYUNI ROLES"
Uyuni server supports the following roles:
.
.TP
\fBorg_admin\fR
Administrative role\. Appears as "Organization Administrator"\. It can administer the entire Uyuni server across all the organisations\.
.
.TP
\fBsatellite_admin\fR
Administrative role\. Appears as "SUSE Manager Administrator"\. It can administer the entire Uyuni server, but only within the given organisation\.
.
.TP
\fBconfig_admin\fR
Appears as "Configuration Administrator" and gives user to configure individual system profiles, channels and certain configuration files\.
.
.TP
\fBchannel_admin\fR
Appears as "Channel Administrator" and gives user ability to add, modify and delete channels\.
.
.TP
\fBsystem_group_admin\fR
Appears as "System Group Administrator" and gives user to access systems section\.
.
.TP
\fBactivation_key_admin\fR
Appears as "Activation Key Administrator" and gives user the control over activation keys, subscriptions etc\.
.
.TP
\fBimage_admin\fR
Appears as "Image Administrator" and is related to OS images administration to build them, store in the registry etc\.
.
.SH "EXAMPLES"
\fBExample: map LDAP roles\fR
.
.P
To map a \fBconfig_admin\fR and \fBchannel_group_admin\fR Uyuni roles to a \fBorganizationalRole\fR object in the LDAP, do the following:
.
.IP "1." 4
Create a role group in the LDAP with the class \fBorganizationalRole\fR\.
.
.IP "2." 4
Add at least one user to that role that is supposed to have it with the attribute \fBroleOccupant\fR\.
.
.IP "3." 4
Create the following configuration in the \fB/etc/rhn/ldapsync\.conf\fR file:
.
.IP "" 0
.
.IP "" 4
.
.nf

  directory:
    roles:
      cn=admins,ou=groups,dc=example,dc=com
        \- config_admin
        \- channel_group_admin
.
.fi
.
.IP "" 0
.
.P
The configuration above will assign a \fBconfig_admin\fR and a \fBchannel_group_admin\fR Uyuni roles to the CN of a role group in LDAP\.
.
.P
\fBExample: map LDAP groups\fR
.
.P
To map a \fBconfig_admin\fR and \fBchannel_group_admin\fR Uyuni roles to a group is very similar to \fBorganizationalRole\fR scenario above, with few differences:
.
.IP "1." 4
Create a group in the LDAP with the class \fBgroupOfNames\fR (POSIX) or \fBgroup\fR (Active Directory)\.
.
.IP "2." 4
Add at least one user to that group with the attribute \fBmember\fR\. NOTE: User attributes should meet the requirements, described in "Requirements" section above\.
.
.IP "3." 4
Create the following configuration in the \fB/etc/rhn/ldapsync\.conf\fR file:
.
.IP "" 0
.
.IP "" 4
.
.nf

   directory:
     groups:
       cn=admins,ou=groups,dc=example,dc=com
         \- config_admin
         \- channel_group_admin
.
.fi
.
.IP "" 0
.
.P
The configuration above will assign a \fBconfig_admin\fR and a \fBchannel_group_admin\fR Uyuni roles to the CN of a group in LDAP\.
.
.P
For more information, look into the configuration file itself and follow the examples there\.
.
.SH "DIAGNOSTICS"
\fBmgr\-ldapsync\fR returns zero on normal operation, non\-zero otherwise\.
.
.SH "AUTHOR"
Bo Maryniuk \fIbo@suse\.de\fR
.
.SH "SEE ALSO"
pam


ldapsync.go
package main

import (
	"fmt"
	"github.com/go-ldap/ldap"
	"github.com/thoas/go-funk"
	"strings"
)

type SearchConfig struct {
	config    *map[string][]string
	filter    string
	attribute string
}

type LDAPSync struct {
	lc           *LDAPCaller
	uc           *UyuniCaller
	cr           *ConfigReader
	ldapusers    []*UyuniUser
	uyuniusers   []*UyuniUser
	allldapusers []*UyuniUser
	roleConfigs  [2]*SearchConfig
}

func NewLDAPSync(cfgpath string) *LDAPSync {
	sync := new(LDAPSync)
	sync.cr = NewConfigReader(cfgpath)
	sync.lc = NewLDAPCaller().
		SetHost(sync.cr.Config().Directory.Host).
		SetPort(sync.cr.Config().Directory.Port).
		SetUser(sync.cr.Config().Directory.User).
		SetPassword(sync.cr.Config().Directory.Password)

	sync.uc = NewUyuniCaller(sync.cr.Config().Spacewalk.Url, !sync.cr.Config().Spacewalk.Checkssl).
		SetUser(sync.cr.Config().Spacewalk.User).
		SetPassword(sync.cr.Config().Spacewalk.Password)
	sync.ldapusers = make([]*UyuniUser, 0)
	sync.uyuniusers = make([]*UyuniUser, 0)
	sync.allldapusers = make([]*UyuniUser, 0)

	sync.roleConfigs = [2]*SearchConfig{
		&SearchConfig{config: &sync.cr.Config().Directory.Roles,
			filter: "(objectClass=organizationalRole)", attribute: "roleOccupant"},
		&SearchConfig{config: &sync.cr.Config().Directory.Groups,
			filter: "(|(objectClass=groupOfNames)(objectClass=group))", attribute: "member"},
	}
	return sync
}

func (sync *LDAPSync) Start() *LDAPSync {
	sync.lc.Connect()

	sync.verifyIgnoredUsers()
	sync.refreshExistingUyuniUsers()
	sync.refreshStagedLDAPUsers()
	sync.refreshAllLDAPUsers()
	sync.refreshUyuniUsersStatus()

	return sync
}

func (sync *LDAPSync) Finish() {
	sync.lc.Disconnect()
}

// Helper function that looks for the same user or at least its ID
func (sync LDAPSync) in(user UyuniUser, users []*UyuniUser) bool {
	for _, u := range users {
		if u.Uid == user.Uid {
			return true
		}
	}
	return false
}

// Match a given user by a DN, compare all metadata.
func (sync LDAPSync) sameAsIn(user *UyuniUser, users []*UyuniUser) (bool, error) {
	for _, u := range users {
		if u.Uid == user.Uid {
			same := u.Email == user.Email
			if same {
				same = u.Name == user.Name
			} else {
				user.accountchanged = true
				log.Debugf("User %s email has been changed from %s to %s", user.Uid, user.Email, u.Email)
			}

			if same {
				same = u.Secondname == user.Secondname
			} else {
				user.accountchanged = true
				log.Debugf("User %s name has been changed from %s to %s", user.Uid, user.Name, u.Name)
			}

			if same {
				same = CompareRoles(user, u)
			} else {
				user.accountchanged = true
				log.Debugf("User %s family name has been changed from %s to %s", user.Uid, user.Secondname, u.Secondname)
			}

			if !same {
				user.roleschanged = true
				log.Debugf("User %s role set has been changed", user.Uid)
			}

			return same, nil
		}
	}

	log.Debugf("Unable to compare user '%s': user not found", user.Uid)

	return false, fmt.Errorf("User UID %s was not found", user.Uid)
}

// Returns a copy of LDAP user by Uyuni user
func (sync *LDAPSync) updateFromLDAPUser(uyuniUser *UyuniUser) {
	for _, ldapUser := range sync.ldapusers {
		if ldapUser.Uid == uyuniUser.Uid {
			uyuniUser.Name, uyuniUser.Secondname, uyuniUser.Email = ldapUser.Name, ldapUser.Secondname, ldapUser.Email
			uyuniUser.FlushRoles()
			for _, role := range ldapUser.GetRoles() {
				uyuniUser.AddRoles(role)
			}
		}
	}
}

// GetDeletedUsers returns an array of users that has been deleted from Uyuni.
// They are both in Uyuni and LDAP, but they are not specified in the LDAP admin-related groups.
func (sync *LDAPSync) GetDeletedUsers() []*UyuniUser {
	var users []*UyuniUser
	for _, user := range sync.allldapusers {
		if !sync.in(*user, sync.ldapusers) && sync.in(*user, sync.uyuniusers) {
			users = append(users, user)
		}
	}

	return users
}

// GetNewUsers returns LDAP users that are not yet in the Uyuni
func (sync *LDAPSync) GetNewUsers() []*UyuniUser {
	var users []*UyuniUser
	for _, user := range sync.uyuniusers {
		if user.IsNew() {
			sync.updateFromLDAPUser(user)
			users = append(users, user)
		}
	}

	return users
}

// GetOutdatedUsers returns LDAP users that are in the Uyuni, but needs refresh
func (sync *LDAPSync) GetOutdatedUsers() []*UyuniUser {
	var users []*UyuniUser
	for _, user := range sync.uyuniusers {
		if !user.IsNew() && user.IsOutdated() {
			users = append(users, user)
		}
	}

	return users
}

// SyncUsers is creating new users in Uyuni by their names and emails.
func (sync *LDAPSync) SyncUsers() []*UyuniUser {
	log.Info("Begin user synchronisation between LDAP and Uyuni server")

	failed := make([]*UyuniUser, 0)
	newUsers := sync.GetNewUsers()
	if len(newUsers) > 0 {
		log.Debugf("Found %d new users", len(newUsers))
		for _, user := range newUsers {
			log.Debugf("New user: %s", user.Uid)
			_, user.Err = sync.uc.Call("user.create", sync.uc.Session(), user.Uid, "", user.Name, user.Secondname, user.Email, 1)

			if !user.IsValid() {
				failed = append(failed, user)
				log.Debugf("Failed to create user %s due to %s", user.Uid, user.Err.Error())
			} else {
				sync.pushUserRolesToUyuni(user)
			}
		}
	}

	existingUsers := sync.GetOutdatedUsers()
	if len(existingUsers) > 0 {
		log.Debugf("Updating %d users", len(existingUsers))
		for _, user := range existingUsers {
			log.Debugf("Update data for user: %s", user.Uid)
			sync.pushUserRolesToUyuni(user)
			sync.pushUserAccountDataToUyuni(user)
		}
	}

	deletedUsers := sync.GetDeletedUsers()
	if len(deletedUsers) > 0 {
		log.Debugf("Deleting removed %d users", len(deletedUsers))
		for _, user := range deletedUsers {
			log.Debugf("Remove user: %s", user.Uid)
			sync.deleteUser(user)
		}
	}

	log.Infof("Added %d new users, updated %d existing users, removed %d users", len(newUsers), len(existingUsers), len(deletedUsers))
	log.Info("End user synchronisation between LDAP and Uyuni server")

	return failed
}

// Remove user from the Uyuni
func (sync *LDAPSync) deleteUser(uyuniUser *UyuniUser) {
	_, err := sync.uc.Call("user.delete", sync.uc.Session(), uyuniUser.Uid)
	if err != nil {
		log.Errorf("Cannot delete users '%s': %s", uyuniUser.Uid, err.Error())
	}
}

// Push account data to Uyuni
func (sync *LDAPSync) pushUserAccountDataToUyuni(user *UyuniUser) {
	_, err := sync.uc.Call("user.setDetails", sync.uc.Session(), user.Uid, map[string]string{
		"first_name": user.Name, "last_name": user.Secondname, "email": user.Email})
	if err != nil {
		log.Errorf("Failed to push user account data for %s: %s", user.Uid, err.Error())
		return
	}

	_, err = sync.uc.Call("user.usePamAuthentication", sync.uc.Session(), user.Uid, 1)
	if err != nil {
		log.Errorf("Failed to push user authentication settings for %s: %s", user.Uid, err.Error())
		return
	}
}

// Sync user roles
func (sync *LDAPSync) pushUserRolesToUyuni(uyuniUser *UyuniUser) {
	// Remove current roles away
	ret, err := sync.uc.Call("user.listRoles", sync.uc.Session(), uyuniUser.Uid)
	if err != nil {
		log.Errorf("Cannot list roles for user '%s': %s", uyuniUser.Uid, err.Error())
		return
	}

	for _, role := range ret.([]interface{}) {
		_, err := sync.uc.Call("user.removeRole", sync.uc.Session(), uyuniUser.Uid, role.(string))
		if err != nil {
			log.Errorf("Cannot remove role '%s': %s", role, err.Error())
		} else {
			log.Debugf("Removed role '%s'", role)
		}
	}

	// Add new roles
	for _, role := range uyuniUser.GetRoles() {
		_, err := sync.uc.Call("user.addRole", sync.uc.Session(), uyuniUser.Uid, role)
		if err != nil {
			log.Errorf("Cannot add role '%s': %s", role, err.Error())
		} else {
			log.Debugf("Added role '%s'", role)
		}
	}
}

// Iterate over possible attribute aliases
func (sync LDAPSync) getAttributes(entry *ldap.Entry, attr ...string) string {
	for _, a := range attr {
		obj := entry.GetAttributeValue(a)
		if obj != "" {
			return obj
		}
	}

	return ""
}

// At least one ignored/frozen user must have org_admin role
func (sync *LDAPSync) verifyIgnoredUsers() {
	valid := false
	for _, uid := range sync.cr.Config().Directory.Frozen {
		res, err := sync.uc.Call("user.listRoles", sync.uc.Session(), uid)
		if err != nil {
			log.Errorf("No users has been found with the UID '%s'", uid)
		} else {
			for _, role := range res.([]interface{}) {
				if role.(string) == "org_admin" {
					valid = true
					goto End
				}
			}
		}
	}
End:
	if !valid {
		log.Fatal("In Uyuni server no actual frozen accounts found with the role 'org_admin'. " +
			"You are risking permanently locking Uyuni server, if you have incorrect LDAP users settings.")
	}
}

// Refresh what users are new and what needs update
func (sync *LDAPSync) refreshUyuniUsersStatus() []*UyuniUser {
	var newusers []*UyuniUser
	for _, user := range sync.ldapusers {
		user.new = !sync.in(*user, sync.uyuniusers)
		if !user.IsNew() {
			isSame, err := sync.sameAsIn(user, sync.uyuniusers)
			if err == nil {
				user.outdated = !isSame
			}
		} else {
			newusers = append(newusers, user.Clone())
		}

		for _, uUuser := range sync.uyuniusers {
			if uUuser.Uid == user.Uid {
				uUuser.outdated = user.outdated
				uUuser.Name = user.Name
				uUuser.Secondname = user.Secondname
				uUuser.Email = user.Email
				uUuser.FlushRoles().AddRoles(user.GetRoles()...)
			}
		}
	}

	sync.uyuniusers = append(sync.uyuniusers, newusers...)

	return sync.uyuniusers
}

// Get all existing users in Uyuni.
func (sync *LDAPSync) refreshExistingUyuniUsers() []*UyuniUser {
	sync.uyuniusers = nil
	res, err := sync.uc.Call("user.listUsers", sync.uc.Session())
	if err != nil {
		log.Fatal(err)
	}
	for _, usrdata := range res.([]interface{}) {
		uid := usrdata.(map[string]interface{})["login"].(string)
		if funk.Contains(sync.cr.Config().Directory.Frozen, uid) {
			continue
		}

		user := NewUyuniUser()
		user.Uid = uid

		res, err = sync.uc.Call("user.getDetails", sync.uc.Session(), user.Uid)
		if err != nil {
			log.Fatal(err)
		}
		userDetails := res.(map[string]interface{})

		user.Email = userDetails["email"].(string)
		user.Name = userDetails["first_name"].(string)
		user.Secondname = userDetails["last_name"].(string)

		// Get user roles
		res, err = sync.uc.Call("user.listRoles", sync.uc.Session(), user.Uid)
		if err != nil {
			log.Fatal(err)
		}

		for _, roleItf := range res.([]interface{}) {
			user.AddRoles(roleItf.(string))
		}

		sync.uyuniusers = append(sync.uyuniusers, user)
	}
	return sync.uyuniusers
}

// Get an attribute name for DN.
// This allows to substitute remapped fields from the configuration, returning
// new remapped name, or keep the original one.
func (sync *LDAPSync) getAttributeNameFor(attr string) string {
	if fieldmap, ext := sync.cr.Config().Directory.Attrmap[sync.cr.Config().Directory.Allusers]; ext {
		nAttr, ext := fieldmap[attr]
		if ext {
			attr = nAttr
		}
	}

	return attr
}

func (sync *LDAPSync) newUserFromDN(dn string) *UyuniUser {
	user := NewUyuniUser()
	request := ldap.NewSearchRequest(dn, ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
		"(objectClass=*)", []string{}, nil)

	entries := sync.lc.Search(request).Entries
	if len(entries) == 1 {
		entry := entries[0]
		user.Dn = entry.DN
		user.Uid = entry.GetAttributeValue(sync.getAttributeNameFor("uid"))
		user.Email = entry.GetAttributeValue(sync.getAttributeNameFor("mail"))

		cn := strings.Split(entry.GetAttributeValue("cn"), " ")
		if len(cn) == 2 {
			user.Name, user.Secondname = cn[0], cn[1]
		} else {
			user.Name = sync.getAttributes(entry, sync.getAttributeNameFor("name"), sync.getAttributeNameFor("givenName"))
			user.Secondname = entry.GetAttributeValue(sync.getAttributeNameFor("sn"))
		}
	} else {
		log.Errorf("DN '%s' matches more or less than one distinct user", dn)
	}

	return user
}

// Get all users from LDAP, regardless are they are meant to be in the Uyuni
func (sync *LDAPSync) refreshAllLDAPUsers() []*UyuniUser {
	sync.allldapusers = nil
	request := ldap.NewSearchRequest(sync.cr.Config().Directory.Allusers,
		ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
		"(objectClass=organizationalPerson)", []string{}, nil)

	for _, entry := range sync.lc.Search(request).Entries {
		sync.allldapusers = append(sync.allldapusers, sync.newUserFromDN(entry.DN))
	}

	return sync.allldapusers
}

// Get existing LDAP users, based on the groups mapping
func (sync *LDAPSync) refreshStagedLDAPUsers() []*UyuniUser {
	sync.ldapusers = nil
	udns := make(map[string]bool)

	// Get all *distinct* user DNs from the "member" attiribute across all the groups
	for _, roleset := range []map[string][]string{sync.cr.Config().Directory.Groups, sync.cr.Config().Directory.Roles} {
		for gdn := range roleset {
			request := ldap.NewSearchRequest(gdn, ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
				"(objectClass=*)", []string{}, nil)
			for _, entry := range sync.lc.Search(request).Entries {
				for _, udn := range append(entry.GetAttributeValues("member"), entry.GetAttributeValues("roleOccupant")...) {
					udns[udn] = true
				}
			}
		}
	}

	// Collect users data
	for udn := range udns {
		user := sync.newUserFromDN(udn)
		if user.Uid != "" && !funk.Contains(sync.cr.Config().Directory.Frozen, user.Uid) {
			sync.updateLDAPUserRoles(user)
			sync.ldapusers = append(sync.ldapusers, user)
		}
	}

	return sync.ldapusers
}

func (sync *LDAPSync) mergeRolesByAttributes(dn string, user *UyuniUser, filter string, attribute string, uyuniRoles []string) {
	req := ldap.NewSearchRequest(dn, ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false, filter, []string{}, nil)
	for _, entry := range sync.lc.Search(req).Entries {
		for _, roleDn := range entry.GetAttributeValues(attribute) {
			if roleDn == user.Dn {
				user.AddRoles(uyuniRoles...)
			}
		}
	}
}

// Get LDAP organizationalRole based on configuration
func (sync *LDAPSync) updateLDAPUserRoles(user *UyuniUser) {
	for _, searchConfig := range sync.roleConfigs {
		for dn, uyuniRoles := range *searchConfig.config {
			sync.mergeRolesByAttributes(dn, user, searchConfig.filter, searchConfig.attribute, uyuniRoles)
		}
	}
}
